### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Functor): refactoring Ran

### PR DESCRIPTION
**Pull request by joelriou**
This PR refactors the definition of the right Kan extension functor (`Functor.ran : (C ⥤ E) ⥤ (D ⥤ E)`) which sends a functor to its right Kan extension along a given functor `F : C ⥤ D`. It used to be defined under the existence of certain colimits. It is now part of the general Kan extension API.

The only significant change in this PR is about showing that the right Kan extension functor associated to a cocontinuous functors between sites sends sheaves to sheaves (this is in the file `CategoryTheory.Sites.CoverLifting`). As this functor is no longer based on the limits API, the proofs had to be changed. In the new proof, we proceed directly with sheaves with values in a category `A` rather than translating the sheaf property in terms of sheaves of types.

This PR is the dual counterpart of #10425 which refactored left Kan extensions.

---

As I had to make a slightly nontrivial change in the file `Condensed.Solid` (currently a leaf file in mathlib), it would be nice to a double-check that the new definitions are mathematically correct.

- [x] depends on: #14261
- [x] depends on: #14256
- [x] depends on: #14258

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joelriou** [comment]
!bench

**Comment by joelriou** [comment]
Note: the variations in the benchmark are completely explained by additions/removals of content in these files.

File: Mathlib/Condensed/Solid.lean [hunk 1]
```diff
@@ -43,11 +43,28 @@ abbrev profiniteFree : Profinite.{u} ⥤ CondensedMod.{u} R :=
...
+right Kan extension of `finFree R : FintypeCat.{u} ⥤ CondensedMod.{u} R`
+along `FintypeCat.toProfinite`. -/
+def profiniteSolidIsPointwiseRightKanExtension :
+    (Functor.RightExtension.mk _ (profiniteSolidCounit R)).IsPointwiseRightKanExtension :=
+  Functor.isPointwiseRightKanExtensionOfIsRightKanExtension _ _
 
 /-- The natural transformation `R[S] ⟶ R[S]^\solid`. -/
 def profiniteSolidification : profiniteFree R ⟶ profiniteSolid.{u} R :=
-  (Ran.equiv FintypeCat.toProfinite (finFree R) (profiniteFree R)).symm (NatTrans.id _)
+  (profiniteSolid R).liftOfIsRightKanExtension (profiniteSolidCounit R) _ (𝟙 _)
```
Comment (dagurtomas) [inline Mathlib/Condensed/Solid.lean:68]: This is all fine, thanks!

File: Mathlib/CategoryTheory/Sites/CoverLifting.lean [hunk 2]
```diff
@@ -92,217 +93,161 @@ theorem isCocontinuous_comp [G.IsCocontinuous J K] [G'.IsCocontinuous K L] :
...
+  intros X S
+  exact ⟨RanIsSheafOfIsCocontinuous.isLimitMultifork ℱ.2
+    (G.op.isPointwiseRightKanExtensionRanCounit ℱ.val) S⟩
 set_option linter.uppercaseLean3 false in
 #align category_theory.Ran_is_sheaf_of_cover_lifting CategoryTheory.ran_isSheaf_of_isCocontinuous
 
 variable (A J)
 
 /-- A cover-lifting functor induces a pushforward functor on categories of sheaves. -/
 def Functor.sheafPushforwardCocontinuous : Sheaf J A ⥤ Sheaf K A where
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/CoverLifting.lean:237]:
```suggestion
/-- A cocontinuous functor induces a pushforward functor on categories of sheaves. -/
def Functor.sheafPushforwardCocontinuous : Sheaf J A ⥤ Sheaf K A where
```

File: Mathlib/CategoryTheory/Sites/CoverLifting.lean [hunk 3]
```diff
@@ -316,92 +261,98 @@ left adjoint to `G.sheafPushforwardCocontinuous A J K`. This adjunction may repl
...
-  homEquiv_counit := by
-    -- Porting note: next line was `ext1`
-    refine Sheaf.Hom.ext _ _ ?_
-    apply (Ran.adjunction A G.op).homEquiv_counit
+noncomputable def sheafAdjunctionCocontinuous :
+    G.sheafPushforwardContinuous A J K ⊣ G.sheafPushforwardCocontinuous A J K :=
+  (G.op.ranAdjunction A).restrictFullyFaithful
+    (fullyFaithfulSheafToPresheaf K A) (fullyFaithfulSheafToPresheaf J A)
+    (G.sheafPushforwardContinuousCompSheafToPresheafIso A J K).symm
+    (G.sheafPushforwardCocontinuousCompSheafToPresheafIso A J K).symm
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/CoverLifting.lean:277]: Nice!

**Review by dagurtomas — APPROVED** [review]
Looks great, the changes in the file about solid modules are mathematically correct.

**Comment by joelriou** [comment]
Thanks @dagurtomas for all the reviews! This is the last step of the refactor!

**Comment by erdOne** [comment]
thanks!
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**